### PR TITLE
feat: Enable MSI for aks cluster

### LINKED ISSUE #4: Enable and configure Managed Identities for the AKS cluster using Terraform
# Summary
Similar to what we do for EKS and are preparing for GKE, we want to define fine grained permissions for any cluster where Jenkins X will be installed.

An example for EKS can be found in the Jenkins-x module, here:

https://github.com/jenkins-x/jx-cloud-provisioners/blob/master/eks/terraform/jenkins-x/irsa.tf

Here we create IAM Policies and IAM Roles that will be used to annotate Kubernetes `Service Accounts` so any pod using these `Service Accounts` can assume the roles.

The idea is to configure the AKS cluster to use `Managed Identities` an have an easy way to define new Service Accounts and permissions.

For now, the `Service Account` being configured with IAM permissions should be:

 * tekton-bot: This is the `Service Account` that the pods running the builds use. The necessary permissions will be defined later.

# Acceptance Criteria
Configure the AKS cluster to use `Managed Identities` and define an easy way to create and configure new `Service Accounts` with the necessary permissions.

### PR DESCRIPTION
**Pull request by helayoty**
In this PR:

- Enable MSI for ask cluster
- install aad-pod-identity helm to assign the user identity we created to the ask
- create role assignments for aks node

Fixes #4

### CODE REVIEW

**Comment by helayoty** [comment]
@chrismellard  Kindly if you could review

**Comment by chrismellard** [comment]
Happy to review. But firstly what's the scope of this PR? It looks like it doesn't directly or entirely address the linked issue. 

As part of #9 I was planning on adding in the Aad-pod stuff as Velero can use it to do fine grained access control in to Azure storage

The linked issue #4 appears to be more to do with linking k8s service accounts with managed identities - I don't believe aad_pod can help out with this though it's probably worth asking over in that project if they can confirm that.

File: local.tf [hunk 1]
```diff
@@ -70,6 +70,15 @@ resource "random_pet" "domain_name" {
   }
 }
 
+resource "random_pet" "msi_name" {
+  prefix    = "msi-aks"
```
Comment (hferentschik) [inline local.tf:74]: I think this goes a bit into the wrong direction. We are using _random_pet_ in `locals.tf` 8 times. This gives us 8 different random pets. IMO we should only generate a single random pet resource and combine it with various suffixes. This way the random pet name can be used across resources to identify resources belonging to the same setup.

Comment (chrismellard) [reply local.tf:74]: These 8 were my doing. We can coalesce on to a single random string and prefix where appropriate. I'll open a small PR in a minute to address

Comment (chrismellard) [reply local.tf:74]: Variable tidy up PR here #11

Comment (helayoty) [reply local.tf:74]: Any chance we can review this PR so I can add more features for aks cluster?

File: modules/cluster/main.tf [hunk 2]
```diff
@@ -23,13 +23,70 @@ resource "azurerm_kubernetes_cluster" "aks" {
 
   role_based_access_control {
     enabled = true
+  }
+}
 
-    azure_active_directory {
```
Comment (chrismellard) [inline modules/cluster/main.tf:27]: Not sure I understand why this is being removed as it is required to enable RBAC integration to AKS? Isn't this tangential to pods running under their own service principals?

File: modules/cluster/main.tf [hunk 3]
```diff
@@ -23,13 +23,70 @@ resource "azurerm_kubernetes_cluster" "aks" {
...
-      managed = true
-    }
+resource "azurerm_user_assigned_identity" "mi" {
+  name                = var.msi_name
+  location            = var.location
+  resource_group_name = var.resource_group_name
+}
+
+provider "helm" {
```
Comment (chrismellard) [inline modules/cluster/main.tf:35]: Providers should be defined once and in the root modules  else we run in to provider conflicts in sub modules

File: variables.tf [hunk 4]
```diff
@@ -46,6 +46,9 @@ variable "apex_domain" {
 variable "domain_name" {
   type = string
 }
+variable "msi_name" {
```
Comment (chrismellard) [inline variables.tf:49]: Can we update the  Readme to cover off what AAD pod identity  is about and the variables it introduces?

**Review by chrismellard — CHANGES_REQUESTED** [review]

File: local.tf (hunk 1 above)
Comment (chrismellard) [reply local.tf:74]: Done! See additional comments

**Comment by chrismellard** [comment]
It would also be good to include a test for these AAD pods -  perhaps a test pod deployed that is attempting to authenticate to blob storage under an AAD pod identity. If we can get  my other PR  #12  merged then this gives you  a good test base to work with.

**Comment by chrismellard** [comment]
Forked this and merged in #34

**Pull request closed without merging**